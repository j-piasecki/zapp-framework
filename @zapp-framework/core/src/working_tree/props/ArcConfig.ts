import { BaseConfigBuilder } from './BaseConfig.js'

export function ArcConfig(id: string) {
  return new ArcConfigBuilder(id)
}

export class ArcConfigBuilder extends BaseConfigBuilder {
  constructor(id: string) {
    super(id)

    this.config.startAngle = 0
    this.config.endAngle = 360
  }

  public lineWidth(width: number) {
    this.config.lineWidth = width
    return this
  }

  public color(color: number) {
    this.config.borderColor = color
    return this
  }

  public startAngle(angle: number) {
    this.config.startAngle = angle
    return this
  }

  public endAngle(angle: number) {
    this.config.endAngle = angle
    return this
  }
}
