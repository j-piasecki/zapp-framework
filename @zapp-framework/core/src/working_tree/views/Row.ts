import { ViewNode } from '../ViewNode.js'
import { WorkingTree } from '../WorkingTree.js'
import { NodeType } from '../../NodeType.js'
import { ConfigBuilder } from '../props/Config.js'

export function Row(configBuilder: ConfigBuilder, body?: () => void) {
  const config = configBuilder.build()
  const current = WorkingTree.current as ViewNode

  const context = WorkingTree.create(current, {
    id: config.id,
    type: NodeType.Row,
    config: config,
    body: body,
  })

  current.children.push(context)

  WorkingTree.withContext(context, body)
}
