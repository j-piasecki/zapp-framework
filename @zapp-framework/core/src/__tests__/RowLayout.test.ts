import { WorkingTree } from '../working_tree/WorkingTree'
import { Stack } from '../working_tree/views/Stack'
import { RowConfig } from '../working_tree/props/RowConfig'
import { Alignment, Arrangement } from '../working_tree/props/types'
import { Config } from '../working_tree/props/BaseConfig'
import { Renderer } from '../renderer/Renderer'
import { DummyViewManager } from '../renderer/DummyViewManager'
import { Row } from '../working_tree/views/Row'
import { setViewManager } from '../renderer/ViewManager'
import { RenderedTree } from '../renderer/RenderedTree'

jest.useFakeTimers()
setViewManager(new DummyViewManager())

function getRenderedTreeString() {
  return JSON.stringify(RenderedTree.current, undefined, 2)
}

afterEach(() => {
  WorkingTree.dropAll()
  jest.setSystemTime(0)
})

test('Children of Row(alignment=Start) get positioned correctly', () => {
  Row(RowConfig('row').width(400).height(400).alignment(Alignment.Start).padding(10), () => {
    Stack(Config('inner.1').width(50).height(50))
    Stack(Config('inner.2').width(80).height(80))
    Stack(Config('inner.3').width(100).height(100))
  })

  WorkingTree.performUpdate()
  Renderer.commit(WorkingTree.root)
  Renderer.render()

  expect(getRenderedTreeString()).toMatchSnapshot()
})

test('Children of Row(alignment=Center) get positioned correctly', () => {
  Row(RowConfig('row').width(400).height(400).alignment(Alignment.Center).padding(10), () => {
    Stack(Config('inner.1').width(50).height(50))
    Stack(Config('inner.2').width(80).height(80))
    Stack(Config('inner.3').width(100).height(100))
  })

  WorkingTree.performUpdate()
  Renderer.commit(WorkingTree.root)
  Renderer.render()

  expect(getRenderedTreeString()).toMatchSnapshot()
})

test('Children of Row(alignment=End) get positioned correctly', () => {
  Row(RowConfig('row').width(400).height(400).alignment(Alignment.End).padding(10), () => {
    Stack(Config('inner.1').width(50).height(50))
    Stack(Config('inner.2').width(80).height(80))
    Stack(Config('inner.3').width(100).height(100))
  })

  WorkingTree.performUpdate()
  Renderer.commit(WorkingTree.root)
  Renderer.render()

  expect(getRenderedTreeString()).toMatchSnapshot()
})

test('Children of Row(arrangement=Start) get positioned correctly', () => {
  Row(RowConfig('row').width(400).height(400).arrangement(Arrangement.Start).padding(10), () => {
    Stack(Config('inner.1').width(50).height(50))
    Stack(Config('inner.2').width(80).height(80))
    Stack(Config('inner.3').width(100).height(100))
  })

  WorkingTree.performUpdate()
  Renderer.commit(WorkingTree.root)
  Renderer.render()

  expect(getRenderedTreeString()).toMatchSnapshot()
})

test('Children of Row(arrangement=Center) get positioned correctly', () => {
  Row(RowConfig('row').width(400).height(400).arrangement(Arrangement.Center).padding(10), () => {
    Stack(Config('inner.1').width(50).height(50))
    Stack(Config('inner.2').width(80).height(80))
    Stack(Config('inner.3').width(100).height(100))
  })

  WorkingTree.performUpdate()
  Renderer.commit(WorkingTree.root)
  Renderer.render()

  expect(getRenderedTreeString()).toMatchSnapshot()
})

test('Children of Row(arrangement=End) get positioned correctly', () => {
  Row(RowConfig('row').width(400).height(400).arrangement(Arrangement.End).padding(10), () => {
    Stack(Config('inner.1').width(50).height(50))
    Stack(Config('inner.2').width(80).height(80))
    Stack(Config('inner.3').width(100).height(100))
  })

  WorkingTree.performUpdate()
  Renderer.commit(WorkingTree.root)
  Renderer.render()

  expect(getRenderedTreeString()).toMatchSnapshot()
})

test('Children of Row(arrangement=SpaceBetween) get positioned correctly', () => {
  Row(
    RowConfig('row').width(400).height(400).arrangement(Arrangement.SpaceBetween).padding(10),
    () => {
      Stack(Config('inner.1').width(50).height(50))
      Stack(Config('inner.2').width(80).height(80))
      Stack(Config('inner.3').width(100).height(100))
    }
  )

  WorkingTree.performUpdate()
  Renderer.commit(WorkingTree.root)
  Renderer.render()

  expect(getRenderedTreeString()).toMatchSnapshot()
})

test('Children of Row(arrangement=SpaceAround) get positioned correctly', () => {
  Row(
    RowConfig('row').width(400).height(400).arrangement(Arrangement.SpaceAround).padding(10),
    () => {
      Stack(Config('inner.1').width(50).height(50))
      Stack(Config('inner.2').width(80).height(80))
      Stack(Config('inner.3').width(100).height(100))
    }
  )

  WorkingTree.performUpdate()
  Renderer.commit(WorkingTree.root)
  Renderer.render()

  expect(getRenderedTreeString()).toMatchSnapshot()
})

test('Children of Row(arrangement=SpaceEvenly) get positioned correctly', () => {
  Row(
    RowConfig('row').width(400).height(400).arrangement(Arrangement.SpaceEvenly).padding(10),
    () => {
      Stack(Config('inner.1').width(50).height(50))
      Stack(Config('inner.2').width(80).height(80))
      Stack(Config('inner.3').width(100).height(100))
    }
  )

  WorkingTree.performUpdate()
  Renderer.commit(WorkingTree.root)
  Renderer.render()

  expect(getRenderedTreeString()).toMatchSnapshot()
})
