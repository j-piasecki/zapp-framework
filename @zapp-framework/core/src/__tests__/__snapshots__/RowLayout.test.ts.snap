// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Children of Row(alignment=Center) get positioned correctly 1`] = `
"{
  \\"id\\": \\"#__root\\",
  \\"type\\": \\"root\\",
  \\"config\\": {
    \\"id\\": \\"#__root\\"
  },
  \\"children\\": [
    {
      \\"id\\": \\"row\\",
      \\"type\\": \\"row\\",
      \\"config\\": {
        \\"id\\": \\"row\\",
        \\"width\\": 400,
        \\"height\\": 400,
        \\"alignment\\": 1,
        \\"padding\\": {
          \\"start\\": 10,
          \\"top\\": 10,
          \\"end\\": 10,
          \\"bottom\\": 10
        }
      },
      \\"children\\": [
        {
          \\"id\\": \\"inner.1\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.1\\",
            \\"width\\": 50,
            \\"height\\": 50
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 50,
            \\"height\\": 50,
            \\"x\\": 10,
            \\"y\\": 175,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        },
        {
          \\"id\\": \\"inner.2\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.2\\",
            \\"width\\": 80,
            \\"height\\": 80
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 80,
            \\"height\\": 80,
            \\"x\\": 60,
            \\"y\\": 160,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        },
        {
          \\"id\\": \\"inner.3\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.3\\",
            \\"width\\": 100,
            \\"height\\": 100
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 100,
            \\"height\\": 100,
            \\"x\\": 140,
            \\"y\\": 150,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        }
      ],
      \\"view\\": null,
      \\"zIndex\\": -1,
      \\"layout\\": {
        \\"width\\": 400,
        \\"height\\": 400,
        \\"x\\": 0,
        \\"y\\": 0,
        \\"measured\\": false,
        \\"widthInferred\\": false,
        \\"heightInferred\\": false
      }
    }
  ],
  \\"view\\": null,
  \\"zIndex\\": -1,
  \\"layout\\": {
    \\"width\\": 400,
    \\"height\\": 400,
    \\"x\\": 0,
    \\"y\\": 0,
    \\"measured\\": true,
    \\"widthInferred\\": false,
    \\"heightInferred\\": false
  }
}"
`;

exports[`Children of Row(alignment=End) get positioned correctly 1`] = `
"{
  \\"id\\": \\"#__root\\",
  \\"type\\": \\"root\\",
  \\"config\\": {
    \\"id\\": \\"#__root\\"
  },
  \\"children\\": [
    {
      \\"id\\": \\"row\\",
      \\"type\\": \\"row\\",
      \\"config\\": {
        \\"id\\": \\"row\\",
        \\"width\\": 400,
        \\"height\\": 400,
        \\"alignment\\": 2,
        \\"padding\\": {
          \\"start\\": 10,
          \\"top\\": 10,
          \\"end\\": 10,
          \\"bottom\\": 10
        }
      },
      \\"children\\": [
        {
          \\"id\\": \\"inner.1\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.1\\",
            \\"width\\": 50,
            \\"height\\": 50
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 50,
            \\"height\\": 50,
            \\"x\\": 10,
            \\"y\\": 340,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        },
        {
          \\"id\\": \\"inner.2\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.2\\",
            \\"width\\": 80,
            \\"height\\": 80
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 80,
            \\"height\\": 80,
            \\"x\\": 60,
            \\"y\\": 310,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        },
        {
          \\"id\\": \\"inner.3\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.3\\",
            \\"width\\": 100,
            \\"height\\": 100
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 100,
            \\"height\\": 100,
            \\"x\\": 140,
            \\"y\\": 290,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        }
      ],
      \\"view\\": null,
      \\"zIndex\\": -1,
      \\"layout\\": {
        \\"width\\": 400,
        \\"height\\": 400,
        \\"x\\": 0,
        \\"y\\": 0,
        \\"measured\\": false,
        \\"widthInferred\\": false,
        \\"heightInferred\\": false
      }
    }
  ],
  \\"view\\": null,
  \\"zIndex\\": -1,
  \\"layout\\": {
    \\"width\\": 400,
    \\"height\\": 400,
    \\"x\\": 0,
    \\"y\\": 0,
    \\"measured\\": true,
    \\"widthInferred\\": false,
    \\"heightInferred\\": false
  }
}"
`;

exports[`Children of Row(alignment=Start) get positioned correctly 1`] = `
"{
  \\"id\\": \\"#__root\\",
  \\"type\\": \\"root\\",
  \\"config\\": {
    \\"id\\": \\"#__root\\"
  },
  \\"children\\": [
    {
      \\"id\\": \\"row\\",
      \\"type\\": \\"row\\",
      \\"config\\": {
        \\"id\\": \\"row\\",
        \\"width\\": 400,
        \\"height\\": 400,
        \\"alignment\\": 0,
        \\"padding\\": {
          \\"start\\": 10,
          \\"top\\": 10,
          \\"end\\": 10,
          \\"bottom\\": 10
        }
      },
      \\"children\\": [
        {
          \\"id\\": \\"inner.1\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.1\\",
            \\"width\\": 50,
            \\"height\\": 50
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 50,
            \\"height\\": 50,
            \\"x\\": 10,
            \\"y\\": 10,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        },
        {
          \\"id\\": \\"inner.2\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.2\\",
            \\"width\\": 80,
            \\"height\\": 80
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 80,
            \\"height\\": 80,
            \\"x\\": 60,
            \\"y\\": 10,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        },
        {
          \\"id\\": \\"inner.3\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.3\\",
            \\"width\\": 100,
            \\"height\\": 100
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 100,
            \\"height\\": 100,
            \\"x\\": 140,
            \\"y\\": 10,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        }
      ],
      \\"view\\": null,
      \\"zIndex\\": -1,
      \\"layout\\": {
        \\"width\\": 400,
        \\"height\\": 400,
        \\"x\\": 0,
        \\"y\\": 0,
        \\"measured\\": false,
        \\"widthInferred\\": false,
        \\"heightInferred\\": false
      }
    }
  ],
  \\"view\\": null,
  \\"zIndex\\": -1,
  \\"layout\\": {
    \\"width\\": 400,
    \\"height\\": 400,
    \\"x\\": 0,
    \\"y\\": 0,
    \\"measured\\": true,
    \\"widthInferred\\": false,
    \\"heightInferred\\": false
  }
}"
`;

exports[`Children of Row(arrangement=Center) get positioned correctly 1`] = `
"{
  \\"id\\": \\"#__root\\",
  \\"type\\": \\"root\\",
  \\"config\\": {
    \\"id\\": \\"#__root\\"
  },
  \\"children\\": [
    {
      \\"id\\": \\"row\\",
      \\"type\\": \\"row\\",
      \\"config\\": {
        \\"id\\": \\"row\\",
        \\"width\\": 400,
        \\"height\\": 400,
        \\"arrangement\\": 4,
        \\"padding\\": {
          \\"start\\": 10,
          \\"top\\": 10,
          \\"end\\": 10,
          \\"bottom\\": 10
        }
      },
      \\"children\\": [
        {
          \\"id\\": \\"inner.1\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.1\\",
            \\"width\\": 50,
            \\"height\\": 50
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 50,
            \\"height\\": 50,
            \\"x\\": 85,
            \\"y\\": 10,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        },
        {
          \\"id\\": \\"inner.2\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.2\\",
            \\"width\\": 80,
            \\"height\\": 80
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 80,
            \\"height\\": 80,
            \\"x\\": 135,
            \\"y\\": 10,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        },
        {
          \\"id\\": \\"inner.3\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.3\\",
            \\"width\\": 100,
            \\"height\\": 100
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 100,
            \\"height\\": 100,
            \\"x\\": 215,
            \\"y\\": 10,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        }
      ],
      \\"view\\": null,
      \\"zIndex\\": -1,
      \\"layout\\": {
        \\"width\\": 400,
        \\"height\\": 400,
        \\"x\\": 0,
        \\"y\\": 0,
        \\"measured\\": false,
        \\"widthInferred\\": false,
        \\"heightInferred\\": false
      }
    }
  ],
  \\"view\\": null,
  \\"zIndex\\": -1,
  \\"layout\\": {
    \\"width\\": 400,
    \\"height\\": 400,
    \\"x\\": 0,
    \\"y\\": 0,
    \\"measured\\": true,
    \\"widthInferred\\": false,
    \\"heightInferred\\": false
  }
}"
`;

exports[`Children of Row(arrangement=End) get positioned correctly 1`] = `
"{
  \\"id\\": \\"#__root\\",
  \\"type\\": \\"root\\",
  \\"config\\": {
    \\"id\\": \\"#__root\\"
  },
  \\"children\\": [
    {
      \\"id\\": \\"row\\",
      \\"type\\": \\"row\\",
      \\"config\\": {
        \\"id\\": \\"row\\",
        \\"width\\": 400,
        \\"height\\": 400,
        \\"arrangement\\": 5,
        \\"padding\\": {
          \\"start\\": 10,
          \\"top\\": 10,
          \\"end\\": 10,
          \\"bottom\\": 10
        }
      },
      \\"children\\": [
        {
          \\"id\\": \\"inner.1\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.1\\",
            \\"width\\": 50,
            \\"height\\": 50
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 50,
            \\"height\\": 50,
            \\"x\\": 160,
            \\"y\\": 10,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        },
        {
          \\"id\\": \\"inner.2\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.2\\",
            \\"width\\": 80,
            \\"height\\": 80
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 80,
            \\"height\\": 80,
            \\"x\\": 210,
            \\"y\\": 10,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        },
        {
          \\"id\\": \\"inner.3\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.3\\",
            \\"width\\": 100,
            \\"height\\": 100
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 100,
            \\"height\\": 100,
            \\"x\\": 290,
            \\"y\\": 10,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        }
      ],
      \\"view\\": null,
      \\"zIndex\\": -1,
      \\"layout\\": {
        \\"width\\": 400,
        \\"height\\": 400,
        \\"x\\": 0,
        \\"y\\": 0,
        \\"measured\\": false,
        \\"widthInferred\\": false,
        \\"heightInferred\\": false
      }
    }
  ],
  \\"view\\": null,
  \\"zIndex\\": -1,
  \\"layout\\": {
    \\"width\\": 400,
    \\"height\\": 400,
    \\"x\\": 0,
    \\"y\\": 0,
    \\"measured\\": true,
    \\"widthInferred\\": false,
    \\"heightInferred\\": false
  }
}"
`;

exports[`Children of Row(arrangement=SpaceAround) get positioned correctly 1`] = `
"{
  \\"id\\": \\"#__root\\",
  \\"type\\": \\"root\\",
  \\"config\\": {
    \\"id\\": \\"#__root\\"
  },
  \\"children\\": [
    {
      \\"id\\": \\"row\\",
      \\"type\\": \\"row\\",
      \\"config\\": {
        \\"id\\": \\"row\\",
        \\"width\\": 400,
        \\"height\\": 400,
        \\"arrangement\\": 1,
        \\"padding\\": {
          \\"start\\": 10,
          \\"top\\": 10,
          \\"end\\": 10,
          \\"bottom\\": 10
        }
      },
      \\"children\\": [
        {
          \\"id\\": \\"inner.1\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.1\\",
            \\"width\\": 50,
            \\"height\\": 50
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 50,
            \\"height\\": 50,
            \\"x\\": 35,
            \\"y\\": 10,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        },
        {
          \\"id\\": \\"inner.2\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.2\\",
            \\"width\\": 80,
            \\"height\\": 80
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 80,
            \\"height\\": 80,
            \\"x\\": 135,
            \\"y\\": 10,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        },
        {
          \\"id\\": \\"inner.3\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.3\\",
            \\"width\\": 100,
            \\"height\\": 100
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 100,
            \\"height\\": 100,
            \\"x\\": 265,
            \\"y\\": 10,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        }
      ],
      \\"view\\": null,
      \\"zIndex\\": -1,
      \\"layout\\": {
        \\"width\\": 400,
        \\"height\\": 400,
        \\"x\\": 0,
        \\"y\\": 0,
        \\"measured\\": false,
        \\"widthInferred\\": false,
        \\"heightInferred\\": false
      }
    }
  ],
  \\"view\\": null,
  \\"zIndex\\": -1,
  \\"layout\\": {
    \\"width\\": 400,
    \\"height\\": 400,
    \\"x\\": 0,
    \\"y\\": 0,
    \\"measured\\": true,
    \\"widthInferred\\": false,
    \\"heightInferred\\": false
  }
}"
`;

exports[`Children of Row(arrangement=SpaceBetween) get positioned correctly 1`] = `
"{
  \\"id\\": \\"#__root\\",
  \\"type\\": \\"root\\",
  \\"config\\": {
    \\"id\\": \\"#__root\\"
  },
  \\"children\\": [
    {
      \\"id\\": \\"row\\",
      \\"type\\": \\"row\\",
      \\"config\\": {
        \\"id\\": \\"row\\",
        \\"width\\": 400,
        \\"height\\": 400,
        \\"arrangement\\": 0,
        \\"padding\\": {
          \\"start\\": 10,
          \\"top\\": 10,
          \\"end\\": 10,
          \\"bottom\\": 10
        }
      },
      \\"children\\": [
        {
          \\"id\\": \\"inner.1\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.1\\",
            \\"width\\": 50,
            \\"height\\": 50
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 50,
            \\"height\\": 50,
            \\"x\\": 10,
            \\"y\\": 10,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        },
        {
          \\"id\\": \\"inner.2\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.2\\",
            \\"width\\": 80,
            \\"height\\": 80
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 80,
            \\"height\\": 80,
            \\"x\\": 135,
            \\"y\\": 10,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        },
        {
          \\"id\\": \\"inner.3\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.3\\",
            \\"width\\": 100,
            \\"height\\": 100
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 100,
            \\"height\\": 100,
            \\"x\\": 290,
            \\"y\\": 10,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        }
      ],
      \\"view\\": null,
      \\"zIndex\\": -1,
      \\"layout\\": {
        \\"width\\": 400,
        \\"height\\": 400,
        \\"x\\": 0,
        \\"y\\": 0,
        \\"measured\\": false,
        \\"widthInferred\\": false,
        \\"heightInferred\\": false
      }
    }
  ],
  \\"view\\": null,
  \\"zIndex\\": -1,
  \\"layout\\": {
    \\"width\\": 400,
    \\"height\\": 400,
    \\"x\\": 0,
    \\"y\\": 0,
    \\"measured\\": true,
    \\"widthInferred\\": false,
    \\"heightInferred\\": false
  }
}"
`;

exports[`Children of Row(arrangement=SpaceEvenly) get positioned correctly 1`] = `
"{
  \\"id\\": \\"#__root\\",
  \\"type\\": \\"root\\",
  \\"config\\": {
    \\"id\\": \\"#__root\\"
  },
  \\"children\\": [
    {
      \\"id\\": \\"row\\",
      \\"type\\": \\"row\\",
      \\"config\\": {
        \\"id\\": \\"row\\",
        \\"width\\": 400,
        \\"height\\": 400,
        \\"arrangement\\": 2,
        \\"padding\\": {
          \\"start\\": 10,
          \\"top\\": 10,
          \\"end\\": 10,
          \\"bottom\\": 10
        }
      },
      \\"children\\": [
        {
          \\"id\\": \\"inner.1\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.1\\",
            \\"width\\": 50,
            \\"height\\": 50
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 50,
            \\"height\\": 50,
            \\"x\\": 47.5,
            \\"y\\": 10,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        },
        {
          \\"id\\": \\"inner.2\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.2\\",
            \\"width\\": 80,
            \\"height\\": 80
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 80,
            \\"height\\": 80,
            \\"x\\": 135,
            \\"y\\": 10,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        },
        {
          \\"id\\": \\"inner.3\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.3\\",
            \\"width\\": 100,
            \\"height\\": 100
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 100,
            \\"height\\": 100,
            \\"x\\": 252.5,
            \\"y\\": 10,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        }
      ],
      \\"view\\": null,
      \\"zIndex\\": -1,
      \\"layout\\": {
        \\"width\\": 400,
        \\"height\\": 400,
        \\"x\\": 0,
        \\"y\\": 0,
        \\"measured\\": false,
        \\"widthInferred\\": false,
        \\"heightInferred\\": false
      }
    }
  ],
  \\"view\\": null,
  \\"zIndex\\": -1,
  \\"layout\\": {
    \\"width\\": 400,
    \\"height\\": 400,
    \\"x\\": 0,
    \\"y\\": 0,
    \\"measured\\": true,
    \\"widthInferred\\": false,
    \\"heightInferred\\": false
  }
}"
`;

exports[`Children of Row(arrangement=Start) get positioned correctly 1`] = `
"{
  \\"id\\": \\"#__root\\",
  \\"type\\": \\"root\\",
  \\"config\\": {
    \\"id\\": \\"#__root\\"
  },
  \\"children\\": [
    {
      \\"id\\": \\"row\\",
      \\"type\\": \\"row\\",
      \\"config\\": {
        \\"id\\": \\"row\\",
        \\"width\\": 400,
        \\"height\\": 400,
        \\"arrangement\\": 3,
        \\"padding\\": {
          \\"start\\": 10,
          \\"top\\": 10,
          \\"end\\": 10,
          \\"bottom\\": 10
        }
      },
      \\"children\\": [
        {
          \\"id\\": \\"inner.1\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.1\\",
            \\"width\\": 50,
            \\"height\\": 50
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 50,
            \\"height\\": 50,
            \\"x\\": 10,
            \\"y\\": 10,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        },
        {
          \\"id\\": \\"inner.2\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.2\\",
            \\"width\\": 80,
            \\"height\\": 80
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 80,
            \\"height\\": 80,
            \\"x\\": 60,
            \\"y\\": 10,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        },
        {
          \\"id\\": \\"inner.3\\",
          \\"type\\": \\"stack\\",
          \\"config\\": {
            \\"id\\": \\"inner.3\\",
            \\"width\\": 100,
            \\"height\\": 100
          },
          \\"children\\": [],
          \\"view\\": null,
          \\"zIndex\\": -1,
          \\"layout\\": {
            \\"width\\": 100,
            \\"height\\": 100,
            \\"x\\": 140,
            \\"y\\": 10,
            \\"measured\\": true,
            \\"widthInferred\\": false,
            \\"heightInferred\\": false
          }
        }
      ],
      \\"view\\": null,
      \\"zIndex\\": -1,
      \\"layout\\": {
        \\"width\\": 400,
        \\"height\\": 400,
        \\"x\\": 0,
        \\"y\\": 0,
        \\"measured\\": false,
        \\"widthInferred\\": false,
        \\"heightInferred\\": false
      }
    }
  ],
  \\"view\\": null,
  \\"zIndex\\": -1,
  \\"layout\\": {
    \\"width\\": 400,
    \\"height\\": 400,
    \\"x\\": 0,
    \\"y\\": 0,
    \\"measured\\": true,
    \\"widthInferred\\": false,
    \\"heightInferred\\": false
  }
}"
`;
